File: rebound_rust/porttest/frequency_test.c
/* frequency_test.c — C reference for the frequency_analysis cross-check.
 * Runs reb_frequency_analysis (MFT, FMFT, FMFT2) on a deterministic
 * three-frequency synthetic signal and dumps the raw IEEE-754 bit
 * patterns of the outputs to frequency_c.txt.
 * Part of the rebound_rs port verification. GPL-3.0-or-later. */
#include "rebound.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

static unsigned long long bits(double x){
    unsigned long long u; memcpy(&u,&x,8); return u;
}

int main(void){
    const size_t ndata = 256;
    const size_t nfreq = 3;
    double* input = malloc(sizeof(double)*2*ndata);
    /* Quasi-periodic signal with three frequencies (rad per sample). */
    const double f1 = 0.30, a1 = 1.00, p1 = 0.40;
    const double f2 = 0.55, a2 = 0.35, p2 = 1.90;
    const double f3 = 0.11, a3 = 0.10, p3 = 5.10;
    for (size_t i=0;i<ndata;i++){
        double t = (double)i;
        input[2*i]   = a1*cos(f1*t+p1) + a2*cos(f2*t+p2) + a3*cos(f3*t+p3);
        input[2*i+1] = a1*sin(f1*t+p1) + a2*sin(f2*t+p2) + a3*sin(f3*t+p3);
    }

    FILE* f = fopen("frequency_c.txt","wb");
    int types[3] = {REB_FREQUENCY_ANALYSIS_MFT, REB_FREQUENCY_ANALYSIS_FMFT, REB_FREQUENCY_ANALYSIS_FMFT2};
    const char* names[3] = {"MFT","FMFT","FMFT2"};
    for (int ti=0; ti<3; ti++){
        double output[9] = {0};
        int ret = reb_frequency_analysis(output, nfreq, 0.05, 1.0, types[ti], input, ndata);
        fprintf(f, "%s ret %d\n", names[ti], ret);
        for (size_t k=0;k<3*nfreq;k++){
            fprintf(f, "%zu %016llx\n", k, bits(output[k]));
        }
    }
    fclose(f);
    free(input);
    printf("frequency_test done\n");
    return 0;
}

File: rebound_rust/examples/frequency_test.rs
//! frequency_test.rs — Rust twin of porttest/frequency_test.c.
//! Part of rebound_rs, GPL-3.0-or-later.
// Clippy waivers. A test/example is its own crate and does not inherit
// the crate root's waivers, so they are repeated here. Same justification:
// this code mirrors the C source's idioms, and applying clippy's
// suggestions would obscure the correspondence that makes the port
// reviewable. See rebound_rust.md section 17.
#![allow(clippy::manual_clamp)] // mirrors the C's explicit min/max tests
#![allow(clippy::neg_cmp_op_on_partial_ord)]
#![allow(clippy::excessive_precision)]
#![allow(clippy::identity_op)]
#![allow(clippy::erasing_op)]
#![allow(clippy::needless_range_loop)]
#![allow(clippy::assign_op_pattern)]
#![allow(clippy::field_reassign_with_default)]
#![allow(clippy::too_many_arguments)]
#![allow(clippy::manual_range_contains)]
#![allow(clippy::manual_memcpy)]
#![allow(clippy::manual_swap)]
#![allow(clippy::manual_div_ceil)]
#![allow(clippy::manual_is_multiple_of)]
#![allow(clippy::misrefactored_assign_op)]
#![allow(clippy::neg_multiply)]
#![allow(clippy::collapsible_if)]
#![allow(clippy::collapsible_else_if)]
#![allow(clippy::needless_late_init)]
#![allow(clippy::while_let_loop)]
#![allow(clippy::unnecessary_cast)]
#![allow(clippy::ptr_arg)]
#![allow(clippy::seek_from_current)]
#![allow(clippy::drop_non_drop)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::approx_constant)]
#![allow(clippy::useless_vec)]
#![allow(clippy::type_complexity)]

use rebound_rs::frequency_analysis::*;
use std::io::Write;

fn bits(x: f64) -> u64 {
    x.to_bits()
}

fn main() {
    let ndata: usize = 256;
    let nfreq: usize = 3;
    let mut input = vec![0.0_f64; 2 * ndata];
    // Quasi-periodic signal with three frequencies (rad per sample).
    let (f1, a1, p1) = (0.30_f64, 1.00_f64, 0.40_f64);
    let (f2, a2, p2) = (0.55_f64, 0.35_f64, 1.90_f64);
    let (f3, a3, p3) = (0.11_f64, 0.10_f64, 5.10_f64);
    for i in 0..ndata {
        let t = i as f64;
        input[2 * i] = a1 * (f1 * t + p1).cos() + a2 * (f2 * t + p2).cos() + a3 * (f3 * t + p3).cos();
        input[2 * i + 1] =
            a1 * (f1 * t + p1).sin() + a2 * (f2 * t + p2).sin() + a3 * (f3 * t + p3).sin();
    }

    let mut f = std::fs::File::create("frequency_rust.txt").unwrap();
    let types = [
        REB_FREQUENCY_ANALYSIS_MFT,
        REB_FREQUENCY_ANALYSIS_FMFT,
        REB_FREQUENCY_ANALYSIS_FMFT2,
    ];
    let names = ["MFT", "FMFT", "FMFT2"];
    for ti in 0..3 {
        let mut output = [0.0_f64; 9];
        let ret = reb_frequency_analysis(&mut output, nfreq, 0.05, 1.0, types[ti], &input, ndata);
        writeln!(f, "{} ret {}", names[ti], ret).unwrap();
        for k in 0..3 * nfreq {
            writeln!(f, "{} {:016x}", k, bits(output[k])).unwrap();
        }
    }
    println!("frequency_test done");
}
